internal/generator/generator_util.go

```diff
@@ -0,0 +1,235 @@
+	// this regex extracts {name} and {definition}.
+	directivesDefBlockExtracter = regexp.MustCompile(`ngx_command_t\s+(\w+)\[\]\s*=\s*{(.*?)};`)
+
+	// Extract one directive definition and attributes from extracted block
+	// { ngx_string({directive_name}),
+	//   {bitmask1|bitmask2|...},
+	//   ... },
+	// this regex extracts {directive_name} and {bitmask1|bitmask2|...}.
+	singleDirectiveExtracter = regexp.MustCompile(`ngx_string\("(.*?)"\).*?,(.*?),`)
+
+	singleLineCommentExtracter = regexp.MustCompile(`//.*`)
+
+	multiLineCommentExtracter = regexp.MustCompile(`/\*[\s\S]*?\*/`)
+)
+
+// Template of support file. A support file contains a map from
+// diective to its bitmask definitions, and a MatchFunc for it.
+//
+//go:embed tmpl/support_file.tmpl
+var supportFileTmplStr string
+
+//nolint:gochecknoglobals
+var supportFileTmpl = template.Must(template.New("supportFile").
+	Funcs(template.FuncMap{"Join": strings.Join}).Parse(supportFileTmplStr))
+
+//nolint:gochecknoglobals
+var ngxBitmaskToGo = map[string]string{
+	"NGX_MAIL_MAIN_CONF":   "ngxMailMainConf",
+	"NGX_STREAM_MAIN_CONF": "ngxStreamMainConf",
+	"NGX_CONF_TAKE1":       "ngxConfTake1",
+	"NGX_STREAM_UPS_CONF":  "ngxStreamUpsConf",
+	"NGX_HTTP_LIF_CONF":    "ngxHTTPLifConf",
+	"NGX_CONF_TAKE2":       "ngxConfTake2",
+	"NGX_HTTP_UPS_CONF":    "ngxHTTPUpsConf",
+	"NGX_CONF_TAKE23":      "ngxConfTake23",
+	"NGX_CONF_TAKE12":      "ngxConfTake12",
+	"NGX_HTTP_MAIN_CONF":   "ngxHTTPMainConf",
+	"NGX_HTTP_LMT_CONF":    "ngxHTTPLmtConf",
+	"NGX_CONF_TAKE1234":    "ngxConfTake1234",
+	"NGX_MAIL_SRV_CONF":    "ngxMailSrvConf",
+	"NGX_CONF_FLAG":        "ngxConfFlag",
+	"NGX_HTTP_SRV_CONF":    "ngxHTTPSrvConf",
+	"NGX_CONF_1MORE":       "ngxConf1More",
+	"NGX_ANY_CONF":         "ngxAnyConf",
+	"NGX_CONF_TAKE123":     "ngxConfTake123",
+	"NGX_MAIN_CONF":        "ngxMainConf",
+	"NGX_CONF_NOARGS":      "ngxConfNoArgs",
+	"NGX_CONF_2MORE":       "ngxConf2More",
+	"NGX_CONF_TAKE3":       "ngxConfTake3",
+	"NGX_HTTP_SIF_CONF":    "ngxHTTPSifConf",
+	"NGX_EVENT_CONF":       "ngxEventConf",
+	"NGX_CONF_BLOCK":       "ngxConfBlock",
+	"NGX_HTTP_LOC_CONF":    "ngxHTTPLocConf",
+	"NGX_STREAM_SRV_CONF":  "ngxStreamSrvConf",
+	"NGX_DIRECT_CONF":      "ngxDirectConf",
+	"NGX_CONF_TAKE13":      "ngxConfTake13",
+	"NGX_CONF_ANY":         "ngxConfAny",
+	"NGX_CONF_TAKE4":       "ngxConfTake4",
+}
+
+//nolint:gochecknoglobals
+var allNgxContexts = map[string]struct{}{
+	"ngxMainConf":       {},
+	"ngxEventConf":      {},
+	"ngxMailMainConf":   {},
+	"ngxMailSrvConf":    {},
+	"ngxStreamMainConf": {},
+	"ngxStreamSrvConf":  {},
+	"ngxStreamUpsConf":  {},
+	"ngxHTTPMainConf":   {},
+	"ngxHTTPSrvConf":    {},
+	"ngxHTTPLocConf":    {},
+	"ngxHTTPUpsConf":    {},
+	"ngxHTTPSifConf":    {},
+	"ngxHTTPLifConf":    {},
+	"ngxHTTPLmtConf":    {},
+	"ngxMgmtMainConf":   {},
+}
+
+//nolint:gochecknoglobals
+var directiveBlock2Context = map[string]string{
+	"ngx_mgmt_block_commands": "ngxMgmtMainConf",
```

Comment:
This is the name of the array of commands defined by the `mgmt` module? What should happen when there is another module that handles context in the same way? What happens if the module implements more than one context?

Maybe this is OK for now as an acceptable limitation.

Reply:
> What should happen when there is another module that handles context in the same way

Say there is a module `MD`, have an array of directive `ngx_MD_block_commands`, all directives here have no context, but actually they should be in block `MD{}` In this case we should add `"ngx_MD_block_commands": MDContext ` into `directiveBlock2Context`, and add `MDContext` into crossplane. I think since `MDGoContext ` is a new context and not in source code we must update crossplane manually for it? 

> What happens if the module implements more than one context?

The assumption I made is: `ngx_mgmt_block_commands` ends with `block_commands` and the "no context directives" in it should actually have a context of `ngx_mgmt_block`(I assume no 2 context in same `block_commands`). If there is another context in this situation we can just add one item into `directiveBlock2Context`?

Reply:
> What happens if the module implements more than one context?

With my approach, I can deal with more than one new context per module as long as we make certain assumptions in the order in which the block and directive arrays are defined. However, that's not a robust approach either. I think any silver bullet code might help until we see more examples in future releases.

Reply:
I think the ultimate solution for flexibility and fewer assumptions may be to pass some value in to allow some custom behavior? Something like:

```sh
generate -context ngxMgmtMainConf  # add to all discovered directives
generate -context usage_report=ngxMgmtMainConf. # add to specific directive
```

It could get long if there are many directives that need to be touched. I suppose you could solve that somehow with some sort of regex selector syntax? Not sure that we would need it yet.

Reply:
> Something like:
generate -context ngxMgmtMainConf  # add to all discovered directives
generate -context usage_report=ngxMgmtMainConf. # add to specific directive

I'm not sure if it should be an option. My caution is when user doesn't pass the `-context` flag, the generated "no mgmt context but is mgmt block directives" seems a bit weird.

> It could get long if there are many directives that need to be touched. I suppose you could solve that somehow with some sort of regex selector syntax?

To the top of my head the pattern for these ngxMgmtMainConf directives is: all directives in `ngx_mgmt_block_commands` should be added a ngxMgmtMainConf. And then it turns back to the assumption of those ending with `block_commands`...

Reply:
I think we probably don't need to expose more API to let the caller of the command configure it any further at this time. It was for sake of argument. I do think the assumption of the way commands are defined and having to hard code that is a little suspect.

When I look at @dareste's solution I think it doesn't make that same assumption, but it does assume a certain order to commands. This might also work for other packages that do not define a context via flag (I think `otel` and `geoip2` do this). Perhaps his solution of capturing the "last command block" is a bit more flexible, although I'm unsure the best way to translate that directive's name into a flag.

Reply:
> When I look at @dareste's solution I think it doesn't make that same assumption

Maybe I made some confusion here lol. I meant I was inspired by his but not exactly same way. I choose this way because at that time I only knew the example of `mgmt` and I thought the pattern "xxx_ block_commands" is reliable.

> This might also work for other packages that do not define a context via flag (I think otel and geoip2 do this).

I see. I didn't notice the cases of `otel` and `geoip2` before. It seems [geoip2](https://github.com/leev/ngx_http_geoip2_module) directives are in `ngxHTTPMainConf`, but otel has some no-context directives in [gExporterCommands](https://github.com/nginxinc/nginx-otel/blob/main/src/http_module.cpp). It seems my former assumption that it should end with `block_commands` doesn't work, but the name of the array  `gExporterCommands` still relate to the context `otel_exporter`. If we want a new context for `otel_exporter` we can add the mapping from `gExporterCommands` to the new context.

I think the problem boils down to depend on "last command block" or "array name". I found both can work currently but both can fail in the future... Since the points we rely on are not "hard rule".

> Perhaps his solution of capturing the "last command block" is a bit more flexible

Could you specify a bit for it? I'm unsure about the flexibilities and appreciate any insights.

Reply:
@dengsh12 `otel` is a good example of new context creation as in the `mgmt` case. 

I think you're right, tracking the array of no-context commands and trying to relate it with existing commands with valid contexts might be more robust than the current regex approaches.

This logic could work:

- Find command arrays.
- Loop the commands in the command array.
- IF the command in the array has both `NGX_CONF_BLOCK` and `NGX_HTTP_*_CONF` flags, and also has a non-standard set handler function (see [Module Config Directives](https://nginx.org/en/docs/dev/development_guide.html#config_directives) for a list of predefined handlers)
  - THEN `findRelatedCommands`

`findRelatedCommands` would contain a non-trivial logic to inspect the found handler function ([`setExporter`](https://github.com/nginxinc/nginx-otel/blob/d3817136395a326dd9626580d81e0c4703fc1463/src/http_module.cpp#L591) as the handler of `otel_exporter`, for instance), explore it and find usage references to command arrays without context flag ([`gExporterCommands`](https://github.com/nginxinc/nginx-otel/blob/main/src/http_module.cpp#L604), following the example). All those found commands would belong to a new context defined by the command name with the custom handler (`otel_exporter`).

Reply:
Cool! `findRelatedCommands` seems the correct way. Searching into the module self defined config handler and figure the relationship should be enough. 
@ornj I'm unsure in crossplane we need to make it such heavy. From another aspect, what we're doing here is to provide a "hidden override". Since we will provide `-filter` and `-override` in next MR, could we just defer the non-context cases into next MR and provide some "module recommended override" to address this? We need to maintain the  "module recommended override", but if a new non-context directive comes we may aslo need to add a new context(even though we have `findRelatedCommands` we may need, since the new context may not be a variable in crossplane )

Reply:
> but if we need to modify the generator code each time we need to support a new module then I think that degrades the value of having the generator a bit.

Nah, only when the module define a context in such a hidden way we need to modify. If we want to add context for it, we must modify the crossplane since the context is new to crossplane. I wonder if we can just utilize the "-override" option in next MR, since we are basically automatically override it.

For example, if a new module `M` defines a no-context directive `m_non_context`, and a block directive `m_block`. The "actual" context of `m_non_context` is `m_block`, in such situation we can't avoid the effort to add a context for `m_block`. The problem boils down to use some mechanism to scan the `M` source code and  figure out the "actual context" in its self-define config handler(it depends on our assumption, the config handler maybe unpredicted but currently reliable), or just utilize the "-override" to handle `m_non_context`.

Reply:
> I wonder if we can just utilize the "-override" option in next MR, since we are basically automatically override it.

👍 I think a later pull-request is fine.

Reply:
I don't think I have anything more. If there's no further feedback from @dareste I think it's ok to resolve.

Reply:
Looks good to me!